Drop support for OpenSSL < 1.1.1 (Poco 1.12.x compile error on systems with OpenSSL 1.0.0)

**Describe the bug**
I get a compile error for Poco 1.12.0 and 1.12.2 on CentOS 7 (with GCC 10) or Ubuntu 16 (both have OpenSSL 1.0.0) in the Crypto component.  Compilation works for 1.11.1, 1.11.2, and 1.11.3.

**To Reproduce**
CentOS 7, Use devtoolset-10, Download Poco 1.12.2 source.  Compile.

**Expected behavior**
I didn't see anything in the change log to suggest that OpenSSL 1.0.0 support had been dropped for Poco 1.12.x or that OpenSSL 1.1.1 is required.  I thought it would build.

**Logs**
```
/opt/rh/devtoolset-10/root/usr/bin/g++  -DPOCO_ENABLE_CPP11 -DPOCO_ENABLE_CPP14 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DPOCO_STATIC -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_REENTRANT -D_THREAD_SAFE -D_XOPEN_SOURCE=500 -I/home/builder/src/poco_container/poco/Crypto/include -I/home/builder/src/poco_container/poco/Crypto/src -I/home/builder/src/poco_container/poco/Foundation/include -O2 -g -DNDEBUG   -std=gnu++14 -MD -MT Crypto/CMakeFiles/Crypto.dir/src/Envelope.cpp.o -MF Crypto/CMakeFiles/Crypto.dir/src/Envelope.cpp.o.d -o Crypto/CMakeFiles/Crypto.dir/src/Envelope.cpp.o -c /home/builder/src/poco_container/poco/Crypto/src/Envelope.cpp
/home/builder/src/poco_container/poco/Crypto/src/Envelope.cpp: In constructor 'Poco::Crypto::Envelope::Envelope(int)':
/home/builder/src/poco_container/poco/Crypto/src/Envelope.cpp:27:8: error: invalid operands of types 'int' and 'void' to binary 'operator!='
   27 |  if (1 != EVP_CIPHER_CTX_init(_pCtx))
      |      ~ ^~ ~~~~~~~~~~~~~~~~~~~~~~~~~~
      |      |                       |
      |      int                     void
```

```
/home/builder/src/poco_container/poco/Crypto/src/EVPCipherImpl.cpp: In constructor 'Poco::Crypto::{anonymous}::EVPEncryptImpl::EVP
EncryptImpl(const EVP_PKEY*)':
/home/builder/src/poco_container/poco/Crypto/src/EVPCipherImpl.cpp:87:31: error: invalid conversion from 'const EVP_PKEY*' {aka 'c
onst evp_pkey_st*'} to 'EVP_PKEY*' {aka 'evp_pkey_st*'} [-fpermissive]
   87 |    _blockSize = EVP_PKEY_size(_pEVP);
      |                               ^~~~~
      |                               |
      |                               const EVP_PKEY* {aka const evp_pkey_st*}
In file included from /home/builder/src/poco_container/poco/Crypto/include/Poco/Crypto/EV
PPKey.h:28,
                 from /home/builder/src/poco_container/poco/Crypto/include/Poco/Crypto/EVPCipherImpl.h:23,
                 from /home/builder/src/poco_container/poco/Crypto/src/EVPCipherImpl.cpp:15:
/usr/include/openssl/evp.h:959:29: note:   initializing argument 1 of 'int EVP_PKEY_size(EVP_PKEY*)'
  959 | int EVP_PKEY_size(EVP_PKEY *pkey);
      |                   ~~~~~~~~~~^~~~
/home/builder/src/poco_container/poco/Crypto/src/EVPCipherImpl.cpp: In constructor 'Poco::Crypto::{anonymous}::EVPDecryptImpl::EVPDecryptImpl(const EVP_PKEY*)':
/home/builder/src/poco_container/poco/Crypto/src/EVPCipherImpl.cpp:200:31: error: invalid conversion from 'const EVP_PKEY*' {aka 'const evp_pkey_st*'} to 'EVP_PKEY*' {aka 'evp_pkey_st*'} [-fpermissive]
  200 |    _blockSize = EVP_PKEY_size(_pEVP);
      |                               ^~~~~
      |                               |
      |                               const EVP_PKEY* {aka const evp_pkey_st*}
In file included from /home/builder/src/poco_container/poco/Crypto/include/Poco/Crypto/EVPPKey.h:28,
                 from /home/builder/src/poco_container/poco/Crypto/include/Poco/Crypto/EVPCipherImpl.h:23,
                 from /home/builder/src/poco_container/poco/Crypto/src/EVPCipherImpl.cpp:15:
/usr/include/openssl/evp.h:959:29: note:   initializing argument 1 of 'int EVP_PKEY_size(EVP_PKEY*)'
  959 | int EVP_PKEY_size(EVP_PKEY *pkey);
      |                   ~~~~~~~~~~^~~~
```
**Screenshots**
NA

**Please add relevant environment information:**
 - CentOS 7
 - POCO Version 1.12.2
 - OpenSSL 1.0.2k-fips

**Additional context**
NA

**EDIT**: Drop support for old OpenSSL

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -110,7 +110,7 @@ if (WIN32 AND CMAKE_VERSION VERSION_LESS  "4.1.0")
 endif()
 
 if (NOT POCO_MINIMAL_BUILD)
-	find_package(OpenSSL)
+	find_package(OpenSSL 1.1.1)
 	find_package(MySQL)
 	find_package(PostgreSQL)
 	find_package(ODBC)
@@ -410,7 +410,7 @@ if(ENABLE_APACHECONNECTOR)
 endif()
 
 if(ENABLE_CRYPTO OR ENABLE_NETSSL OR ENABLE_JWT)
-	find_package(OpenSSL REQUIRED)
+	find_package(OpenSSL 1.1.1 REQUIRED)
 endif()
 
 if(ENABLE_DATA_MYSQL)

--- a/Crypto/include/Poco/Crypto/Crypto.h
+++ b/Crypto/include/Poco/Crypto/Crypto.h
@@ -45,8 +45,8 @@
 #endif
 
 
-#if OPENSSL_VERSION_NUMBER < 0x10000000L
-#error "OpenSSL version too old. At least OpenSSL 1.0.0 is required."
+#if OPENSSL_VERSION_NUMBER < 0x10101000L
+#error "OpenSSL version too old. At least OpenSSL 1.1.1 is required."
 #endif
 
 

--- a/Crypto/include/Poco/Crypto/EVPPKey.h
+++ b/Crypto/include/Poco/Crypto/EVPPKey.h
@@ -65,8 +65,6 @@ public:
 	EVPPKey(const PKCS12Container& cert);
 		/// Constructs EVPPKey from the given container.
 
-#if OPENSSL_VERSION_NUMBER >= 0x10000000L
-
 	EVPPKey(int type, int param);
 		/// Creates the EVPPKey.
 		/// Creates a new public/private keypair using the given parameters.
@@ -79,10 +77,6 @@ public:
 		/// Parameters:
 		///   - for EVP_PKEY_RSA: key length in bits
 		///   - for EVP_PKEY_EC: curve NID
-		///
-		/// This constructor is not available for OpenSSL version < 1.0.0
-
-#endif // OPENSSL_VERSION_NUMBER >= 0x10000000L
 
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 	explicit EVPPKey(const std::vector<unsigned char>* publicKey, const std::vector<unsigned char>* privateKey, unsigned long exponent, int type);

--- a/Crypto/include/Poco/Crypto/OpenSSLInitializer.h
+++ b/Crypto/include/Poco/Crypto/OpenSSLInitializer.h
@@ -19,25 +19,12 @@
 
 
 #include "Poco/Crypto/Crypto.h"
-#include "Poco/Mutex.h"
 #include "Poco/AtomicCounter.h"
 #include <openssl/crypto.h>
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 #include <openssl/provider.h>
 #include <atomic>
 #endif
-#if defined(OPENSSL_FIPS) && OPENSSL_VERSION_NUMBER < 0x010001000L
-#include <openssl/fips.h>
-#endif
-
-
-extern "C"
-{
-	struct CRYPTO_dynlock_value
-	{
-		Poco::FastMutex _mutex;
-	};
-}
 
 
 namespace Poco {
@@ -72,27 +59,8 @@ public:
 	static bool haveLegacyProvider();
 		/// Returns true if the OpenSSL legacy provider is available, otherwise false.
 
-protected:
-	enum
-	{
-		SEEDSIZE = 256
-	};
-
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-	// OpenSSL multithreading support
-	static void lock(int mode, int n, const char* file, int line);
-	static unsigned long id();
-	static struct CRYPTO_dynlock_value* dynlockCreate(const char* file, int line);
-	static void dynlock(int mode, struct CRYPTO_dynlock_value* lock, const char* file, int line);
-	static void dynlockDestroy(struct CRYPTO_dynlock_value* lock, const char* file, int line);
-#endif
-
 private:
 	static Poco::AtomicCounter _rc;
-
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-	static Poco::FastMutex* _mutexes;
-#endif
 
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 	static OSSL_PROVIDER* _defaultProvider;

--- a/Crypto/src/CipherImpl.cpp
+++ b/Crypto/src/CipherImpl.cpp
@@ -77,11 +77,7 @@ namespace
 
 	private:
 		const EVP_CIPHER* _pCipher;
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		EVP_CIPHER_CTX*   _pContext;
-#else
-		EVP_CIPHER_CTX    _context;
-#endif
 		ByteVec           _key;
 		ByteVec           _iv;
 	};
@@ -96,7 +92,6 @@ namespace
 		_key(key),
 		_iv(iv)
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		_pContext = EVP_CIPHER_CTX_new();
 		if (!_pContext) throwError();
 		int rc = EVP_CipherInit(
@@ -105,87 +100,49 @@ namespace
 			&_key[0],
 			_iv.empty() ? nullptr : &_iv[0],
 			(dir == DIR_ENCRYPT) ? 1 : 0);
-#else
-		int rc = EVP_CipherInit(
-			&_context,
-			_pCipher,
-			&_key[0],
-			_iv.empty() ? 0 : &_iv[0],
-			(dir == DIR_ENCRYPT) ? 1 : 0);
-#endif
 		if (rc == 0) throwError();
 
-#if OPENSSL_VERSION_NUMBER >= 0x10001000L
 		if (static_cast<int>(_iv.size()) != EVP_CIPHER_iv_length(_pCipher) && EVP_CIPHER_mode(_pCipher) == EVP_CIPH_GCM_MODE)
 		{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
-			int rc = EVP_CIPHER_CTX_ctrl(_pContext, EVP_CTRL_GCM_SET_IVLEN, static_cast<int>(_iv.size()), nullptr);
-#else
-			int rc = EVP_CIPHER_CTX_ctrl(&_context, EVP_CTRL_GCM_SET_IVLEN, static_cast<int>(_iv.size()), nullptr);
-#endif
+			rc = EVP_CIPHER_CTX_ctrl(_pContext, EVP_CTRL_GCM_SET_IVLEN, static_cast<int>(_iv.size()), nullptr);
 			if (rc == 0) throwError();
 		}
-#endif
 	}
 
 
 	CryptoTransformImpl::~CryptoTransformImpl()
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		EVP_CIPHER_CTX_reset(_pContext);
 		EVP_CIPHER_CTX_free(_pContext);
-#else
-		EVP_CIPHER_CTX_cleanup(&_context);
-#endif
 	}
 
 
 	std::size_t CryptoTransformImpl::blockSize() const
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		return EVP_CIPHER_CTX_block_size(_pContext);
-#else
-		return EVP_CIPHER_CTX_block_size(&_context);
-#endif
 	}
 
 
 	int CryptoTransformImpl::setPadding(int padding)
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		return EVP_CIPHER_CTX_set_padding(_pContext, padding);
-#else
-		return EVP_CIPHER_CTX_set_padding(&_context, padding);
-#endif
 	}
 
 
 	std::string CryptoTransformImpl::getTag(std::size_t tagSize)
 	{
 		std::string tag;
-#if OPENSSL_VERSION_NUMBER >= 0x10001000L
 		Poco::Buffer<char> buffer(tagSize);
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		int rc = EVP_CIPHER_CTX_ctrl(_pContext, EVP_CTRL_GCM_GET_TAG, static_cast<int>(tagSize), buffer.begin());
-#else
-		int rc = EVP_CIPHER_CTX_ctrl(&_context, EVP_CTRL_GCM_GET_TAG, static_cast<int>(tagSize), buffer.begin());
-#endif
 		if (rc == 0) throwError();
 		tag.assign(buffer.begin(), tagSize);
-#endif
 		return tag;
 	}
 
 
 	void CryptoTransformImpl::setTag(const std::string& tag)
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		int rc = EVP_CIPHER_CTX_ctrl(_pContext, EVP_CTRL_GCM_SET_TAG, static_cast<int>(tag.size()), const_cast<char*>(tag.data()));
-#elif OPENSSL_VERSION_NUMBER >= 0x10001000L
-		int rc = EVP_CIPHER_CTX_ctrl(&_context, EVP_CTRL_GCM_SET_TAG, static_cast<int>(tag.size()), const_cast<char*>(tag.data()));
-#else
-		int rc = 0;
-#endif
 		if (rc == 0) throwError();
 	}
 
@@ -199,21 +156,12 @@ namespace
 		poco_assert (outputLength >= (inputLength + static_cast<std::streamsize>(blockSize()) - 1));
 
 		int outLen = static_cast<int>(outputLength);
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		int rc = EVP_CipherUpdate(
 			_pContext,
 			output,
 			&outLen,
 			input,
 			static_cast<int>(inputLength));
-#else
-		int rc = EVP_CipherUpdate(
-			&_context,
-			output,
-			&outLen,
-			input,
-			static_cast<int>(inputLength));
-#endif
 		if (rc == 0)
 			throwError();
 
@@ -230,13 +178,9 @@ namespace
 		int len = static_cast<int>(length);
 
 		// Use the '_ex' version that does not perform implicit cleanup since we
-		// will call EVP_CIPHER_CTX_cleanup() from the dtor as there is no
+		// will call EVP_CIPHER_CTX_reset() from the dtor as there is no
 		// guarantee that finalize() will be called if an error occurred.
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		int rc = EVP_CipherFinal_ex(_pContext, output, &len);
-#else
-		int rc = EVP_CipherFinal_ex(&_context, output, &len);
-#endif
 
 		if (rc == 0)
 			throwError();

--- a/Crypto/src/CipherKeyImpl.cpp
+++ b/Crypto/src/CipherKeyImpl.cpp
@@ -134,7 +134,6 @@ CipherKeyImpl::Mode CipherKeyImpl::mode() const
 	case EVP_CIPH_OFB_MODE:
 		return MODE_OFB;
 
-#if OPENSSL_VERSION_NUMBER >= 0x10001000L
 	case EVP_CIPH_CTR_MODE:
 		return MODE_CTR;
 
@@ -143,7 +142,6 @@ CipherKeyImpl::Mode CipherKeyImpl::mode() const
 
 	case EVP_CIPH_CCM_MODE:
 		return MODE_CCM;
-#endif
 	}
 	throw Poco::IllegalStateException("Unexpected value of EVP_CIPHER_mode()");
 }

--- a/Crypto/src/EVPPKey.cpp
+++ b/Crypto/src/EVPPKey.cpp
@@ -172,7 +172,6 @@ EVPPKey::EVPPKey(const std::vector<unsigned char>* public_key, const std::vector
 
 #endif
 
-#if OPENSSL_VERSION_NUMBER >= 0x10000000L
 
 EVPPKey::EVPPKey(int type, int param) : _pEVPPKey(nullptr)
 {
@@ -231,8 +230,6 @@ EVPPKey::EVPPKey(int type, int param) : _pEVPPKey(nullptr)
 	EVP_PKEY_CTX_free(pCtx);
 	checkType();
 }
-
-#endif // OPENSSL_VERSION_NUMBER >= 0x10000000L
 
 
 EVPPKey::EVPPKey(EVP_PKEY* pEVPPKey) : _pEVPPKey(nullptr)

--- a/Crypto/src/OpenSSLInitializer.cpp
+++ b/Crypto/src/OpenSSLInitializer.cpp
@@ -14,10 +14,7 @@
 
 #include "Poco/Crypto/OpenSSLInitializer.h"
 #include "Poco/Crypto/CryptoException.h"
-#include "Poco/RandomStream.h"
-#include "Poco/Thread.h"
 #include <openssl/ssl.h>
-#include <openssl/rand.h>
 #include <openssl/crypto.h>
 #include <openssl/err.h>
 #include <openssl/conf.h>
@@ -35,10 +32,6 @@
 	#endif
 	#pragma message (OPENSSL_VERSION_TEXT POCO_INTERNAL_OPENSSL_BUILD)
 #endif
-
-
-using Poco::RandomInputStream;
-using Poco::Thread;
 
 
 #if defined(_MSC_VER) && !defined(_DLL) && defined(POCO_INTERNAL_OPENSSL_MSVC_VER)
@@ -60,10 +53,6 @@ namespace Crypto {
 
 
 Poco::AtomicCounter OpenSSLInitializer::_rc;
-
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-Poco::FastMutex* OpenSSLInitializer::_mutexes(0);
-#endif
 
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 OSSL_PROVIDER* OpenSSLInitializer::_defaultProvider(nullptr);
@@ -94,39 +83,7 @@ void OpenSSLInitializer::initialize()
 {
 	if (++_rc == 1)
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		CONF_modules_load(nullptr, nullptr, 0);
-#else
-		OPENSSL_config(nullptr);
-#endif
-
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-		SSL_library_init();
-		SSL_load_error_strings();
-		OpenSSL_add_all_algorithms();
-
-		int nMutexes = CRYPTO_num_locks();
-		_mutexes = new Poco::FastMutex[nMutexes];
-		CRYPTO_set_locking_callback(&OpenSSLInitializer::lock);
-#ifndef POCO_OS_FAMILY_WINDOWS
-// Not needed on Windows (see SF #110: random unhandled exceptions when linking with ssl).
-// https://sourceforge.net/p/poco/bugs/110/
-//
-// From http://www.openssl.org/docs/crypto/threads.html :
-// "If the application does not register such a callback using CRYPTO_THREADID_set_callback(),
-//  then a default implementation is used - on Windows and BeOS this uses the system's
-//  default thread identifying APIs"
-		CRYPTO_set_id_callback(&OpenSSLInitializer::id);
-#endif
-		CRYPTO_set_dynlock_create_callback(&OpenSSLInitializer::dynlockCreate);
-		CRYPTO_set_dynlock_lock_callback(&OpenSSLInitializer::dynlock);
-		CRYPTO_set_dynlock_destroy_callback(&OpenSSLInitializer::dynlockDestroy);
-
-		char seed[SEEDSIZE];
-		RandomInputStream rnd;
-		rnd.read(seed, sizeof(seed));
-		RAND_seed(seed, SEEDSIZE);
-#endif
 
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 		if (!_defaultProvider)
@@ -148,64 +105,9 @@ void OpenSSLInitializer::uninitialize()
 {
 	if (--_rc == 0)
 	{
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-		EVP_cleanup();
-		ERR_free_strings();
-		CRYPTO_set_locking_callback(0);
-#ifndef POCO_OS_FAMILY_WINDOWS
-		CRYPTO_set_id_callback(0);
-#endif
-		delete [] _mutexes;
-#endif
+		// OpenSSL 1.1.0+ handles cleanup automatically
 	}
 }
-
-
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-
-
-void OpenSSLInitializer::lock(int mode, int n, const char* file, int line)
-{
-	if (mode & CRYPTO_LOCK)
-		_mutexes[n].lock();
-	else
-		_mutexes[n].unlock();
-}
-
-
-unsigned long OpenSSLInitializer::id()
-{
-	// Note: we use an old-style C cast here because
-	// neither static_cast<> nor reinterpret_cast<>
-	// work uniformly across all platforms.
-	return (unsigned long) Poco::Thread::currentTid();
-}
-
-
-struct CRYPTO_dynlock_value* OpenSSLInitializer::dynlockCreate(const char* file, int line)
-{
-	return new CRYPTO_dynlock_value;
-}
-
-
-void OpenSSLInitializer::dynlock(int mode, struct CRYPTO_dynlock_value* lock, const char* file, int line)
-{
-	poco_check_ptr (lock);
-
-	if (mode & CRYPTO_LOCK)
-		lock->_mutex.lock();
-	else
-		lock->_mutex.unlock();
-}
-
-
-void OpenSSLInitializer::dynlockDestroy(struct CRYPTO_dynlock_value* lock, const char* file, int line)
-{
-	delete lock;
-}
-
-
-#endif // OPENSSL_VERSION_NUMBER < 0x10100000L
 
 
 void initializeCrypto()

--- a/Crypto/src/RSAKeyImpl.cpp
+++ b/Crypto/src/RSAKeyImpl.cpp
@@ -218,43 +218,31 @@ int RSAKeyImpl::size() const
 
 RSAKeyImpl::ByteVec RSAKeyImpl::modulus() const
 {
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 	const BIGNUM* n = nullptr;
 	const BIGNUM* e = nullptr;
 	const BIGNUM* d = nullptr;
 	RSA_get0_key(_pRSA, &n, &e, &d);
 	return convertToByteVec(n);
-#else
-	return convertToByteVec(_pRSA->n);
-#endif
 }
 
 
 RSAKeyImpl::ByteVec RSAKeyImpl::encryptionExponent() const
 {
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 	const BIGNUM* n = nullptr;
 	const BIGNUM* e = nullptr;
 	const BIGNUM* d = nullptr;
 	RSA_get0_key(_pRSA, &n, &e, &d);
 	return convertToByteVec(e);
-#else
-	return convertToByteVec(_pRSA->e);
-#endif
 }
 
 
 RSAKeyImpl::ByteVec RSAKeyImpl::decryptionExponent() const
 {
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 	const BIGNUM* n = nullptr;
 	const BIGNUM* e = nullptr;
 	const BIGNUM* d = nullptr;
 	RSA_get0_key(_pRSA, &n, &e, &d);
 	return convertToByteVec(d);
-#else
-	return convertToByteVec(_pRSA->d);
-#endif
 }
 
 

--- a/Crypto/src/X509Certificate.cpp
+++ b/Crypto/src/X509Certificate.cpp
@@ -29,13 +29,6 @@
 #include <openssl/bn.h>
 
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-#define ASN1_STRING_get0_data ASN1_STRING_data
-#define X509_get0_notBefore X509_get_notBefore
-#define X509_get0_notAfter X509_get_notAfter
-#endif
-
-
 namespace Poco {
 namespace Crypto {
 
@@ -68,11 +61,7 @@ X509Certificate::X509Certificate(X509* pCert, bool shared):
 
 	if (shared)
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		X509_up_ref(_pCert);
-#else
-		_pCert->references++;
-#endif
 	}
 
 	init();
@@ -385,14 +374,7 @@ bool X509Certificate::equals(const X509Certificate& otherCertificate) const
 
 std::string X509Certificate::signatureAlgorithm() const
 {
-	int sigNID = NID_undef;
-
-#if (OPENSSL_VERSION_NUMBER >=  0x1010000fL)
-	sigNID = X509_get_signature_nid(_pCert);
-#else
-	poco_check_ptr(_pCert->sig_alg);
-	sigNID = OBJ_obj2nid(_pCert->sig_alg->algorithm);
-#endif
+	int sigNID = X509_get_signature_nid(_pCert);
 
 	if (sigNID != NID_undef)
 	{

--- a/NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h
+++ b/NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h
@@ -23,6 +23,7 @@
 #include "Poco/Net/Context.h"
 #include "Poco/Net/X509Certificate.h"
 #include "Poco/Net/Session.h"
+#include "Poco/Mutex.h"
 #include <openssl/bio.h>
 #include <openssl/ssl.h>
 

--- a/NetSSL_OpenSSL/src/Context.cpp
+++ b/NetSSL_OpenSSL/src/Context.cpp
@@ -209,13 +209,9 @@ void Context::init(const Params& params)
 
 		if (!isForServerUse() && params.ocspStaplingVerification)
 		{
-#if OPENSSL_VERSION_NUMBER >= 0x10001000L
 			_ocspStaplingResponseVerification = true;
 			SSL_CTX_set_tlsext_status_cb(_pSSLContext, &SSLManager::verifyOCSPResponseCallback);
 			SSL_CTX_set_tlsext_status_arg(_pSSLContext, this);
-#else
-			throw SSLContextException("OCSP Stapling is not supported by this OpenSSL version");
-#endif
 		}
 
 		initDH(params.dhGroup, params.dhParamsFile);
@@ -231,9 +227,7 @@ void Context::init(const Params& params)
 
 void Context::setSecurityLevel(SecurityLevel level)
 {
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 	SSL_CTX_set_security_level(_pSSLContext, static_cast<int>(level));
-#endif
 }
 
 void Context::ignoreUnexpectedEof(bool flag)
@@ -459,7 +453,6 @@ void Context::disableProtocols(int protocols)
 
 void Context::requireMinimumProtocol(Protocols protocol)
 {
-#if OPENSSL_VERSION_NUMBER >= 0x10101000L
 	int version = 0;
 	switch (protocol)
 	{
@@ -486,43 +479,6 @@ void Context::requireMinimumProtocol(Protocols protocol)
 		unsigned long err = ERR_get_error();
 		throw SSLException("Cannot set minimum supported version on SSL_CTX object", ERR_error_string(err, nullptr));
 	}
-
-#else
-
-	switch (protocol)
-	{
-	case PROTO_SSLV2:
-		throw Poco::InvalidArgumentException("SSLv2 is no longer supported");
-
-	case PROTO_SSLV3:
-		throw Poco::InvalidArgumentException("SSLv3 is no longer supported");
-		break;
-
-	case PROTO_TLSV1:
-		disableProtocols(PROTO_SSLV2 | PROTO_SSLV3);
-		break;
-
-	case PROTO_TLSV1_1:
-#if defined(SSL_OP_NO_TLSv1_1) && !defined(OPENSSL_NO_TLS1)
-		disableProtocols(PROTO_SSLV2 | PROTO_SSLV3 | PROTO_TLSV1);
-#else
-		throw Poco::InvalidArgumentException("TLSv1.1 is not supported by the available OpenSSL library");
-#endif
-		break;
-
-	case PROTO_TLSV1_2:
-#if defined(SSL_OP_NO_TLSv1_2) && !defined(OPENSSL_NO_TLS1)
-		disableProtocols(PROTO_SSLV2 | PROTO_SSLV3 | PROTO_TLSV1 | PROTO_TLSV1_1);
-#else
-		throw Poco::InvalidArgumentException("TLSv1.2 is not supported by the available OpenSSL library");
-#endif
-		break;
-
-	case PROTO_TLSV1_3:
-		throw Poco::InvalidArgumentException("TLSv1.3 is not supported by the available OpenSSL library");
-		break;
-	}
-#endif
 }
 
 
@@ -546,11 +502,7 @@ void Context::createSSLContext()
 
 	if (SSLManager::isFIPSEnabled())
 	{
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 		_pSSLContext = SSL_CTX_new(TLS_method());
-#else
-		_pSSLContext = SSL_CTX_new(TLSv1_method());
-#endif
 	}
 	else
 	{
@@ -558,99 +510,55 @@ void Context::createSSLContext()
 		{
 		case CLIENT_USE:
 		case TLS_CLIENT_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_client_method());
 			minTLSVersion = TLS1_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(SSLv23_client_method());
-#endif
 			break;
 
 		case SERVER_USE:
 		case TLS_SERVER_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_server_method());
 			minTLSVersion = TLS1_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(SSLv23_server_method());
-#endif
 			break;
 
 		case TLSV1_CLIENT_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_client_method());
 			minTLSVersion = TLS1_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(TLSv1_client_method());
-#endif
 			break;
 
 		case TLSV1_SERVER_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_server_method());
 			minTLSVersion = TLS1_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(TLSv1_server_method());
-#endif
 			break;
 
-#if defined(SSL_OP_NO_TLSv1_1) && !defined(OPENSSL_NO_TLS1)
-/* SSL_OP_NO_TLSv1_1 is defined in ssl.h if the library version supports TLSv1.1.
- * OPENSSL_NO_TLS1 is defined in opensslconf.h or on the compiler command line
- * if TLS1.x was removed at OpenSSL library build time via Configure options.
- */
+#if !defined(OPENSSL_NO_TLS1)
 		case TLSV1_1_CLIENT_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_client_method());
 			minTLSVersion = TLS1_1_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(TLSv1_1_client_method());
-#endif
 			break;
 
 		case TLSV1_1_SERVER_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_server_method());
 			minTLSVersion = TLS1_1_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(TLSv1_1_server_method());
-#endif
 			break;
-#endif
 
-#if defined(SSL_OP_NO_TLSv1_2) && !defined(OPENSSL_NO_TLS1)
 		case TLSV1_2_CLIENT_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_client_method());
 			minTLSVersion = TLS1_2_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(TLSv1_2_client_method());
-#endif
 			break;
 
 		case TLSV1_2_SERVER_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 			_pSSLContext = SSL_CTX_new(TLS_server_method());
 			minTLSVersion = TLS1_2_VERSION;
-#else
-			_pSSLContext = SSL_CTX_new(TLSv1_2_server_method());
-#endif
 			break;
-#endif
 
-#if defined(SSL_OP_NO_TLSv1_3) && !defined(OPENSSL_NO_TLS1)
 		case TLSV1_3_CLIENT_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10101000L
 			_pSSLContext = SSL_CTX_new(TLS_client_method());
 			minTLSVersion = TLS1_3_VERSION;
-#endif
 			break;
 
 		case TLSV1_3_SERVER_USE:
-#if OPENSSL_VERSION_NUMBER >= 0x10101000L
 			_pSSLContext = SSL_CTX_new(TLS_server_method());
 			minTLSVersion = TLS1_3_VERSION;
-#endif
 			break;
 #endif
 
@@ -664,7 +572,6 @@ void Context::createSSLContext()
 		throw SSLException("Cannot create SSL_CTX object", ERR_error_string(err, nullptr));
 	}
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
 	if (minTLSVersion)
 	{
 		if (!SSL_CTX_set_min_proto_version(_pSSLContext, minTLSVersion))
@@ -675,7 +582,6 @@ void Context::createSSLContext()
 			throw SSLException("Cannot set minimum supported version on SSL_CTX object", ERR_error_string(err, nullptr));
 		}
 	}
-#endif
 
 	SSL_CTX_set_default_passwd_cb(_pSSLContext, &SSLManager::privateKeyPassphraseCallback);
 	Utility::clearErrorStack();
@@ -1005,47 +911,12 @@ void Context::initDH(KeyDHGroup keyDHGroup, const std::string& dhParamsFile)
 void Context::initECDH(const std::string& curve)
 {
 #ifndef OPENSSL_NO_ECDH
-#if OPENSSL_VERSION_NUMBER >= 0x1000200fL
-	const std::string groups(curve.empty() ?
- #if   OPENSSL_VERSION_NUMBER >= 0x1010100fL
-				   "X448:X25519:P-521:P-384:P-256"
- #elif OPENSSL_VERSION_NUMBER >= 0x1010000fL
-	// while OpenSSL 1.1.0 didn't support Ed25519 (EdDSA using Curve25519),
-	// it did support X25519 (ECDH using Curve25516).
-				   "X25519:P-521:P-384:P-256"
- #else
-				   "P-521:P-384:P-256"
- #endif
-				   : curve);
+	const std::string groups(curve.empty() ? "X448:X25519:P-521:P-384:P-256" : curve);
 	if (SSL_CTX_set1_curves_list(_pSSLContext, groups.c_str()) == 0)
 	{
 		throw SSLContextException("Cannot set ECDH groups", groups);
 	}
 	SSL_CTX_set_options(_pSSLContext, SSL_OP_SINGLE_ECDH_USE);
- #else
-	int nid = 0;
-	if (!curve.empty())
-	{
-		nid = OBJ_sn2nid(curve.c_str());
-	}
-	else
-	{
-		nid = OBJ_sn2nid("prime256v1");
-	}
-	if (nid == 0)
-	{
-		throw SSLContextException("Unknown ECDH curve name", curve);
-	}
-
-	EC_KEY* ecdh = EC_KEY_new_by_curve_name(nid);
-	if (!ecdh)
-	{
-		throw SSLContextException("Cannot create ECDH curve");
-	}
-	SSL_CTX_set_tmp_ecdh(_pSSLContext, ecdh);
-	SSL_CTX_set_options(_pSSLContext, SSL_OP_SINGLE_ECDH_USE);
-	EC_KEY_free(ecdh);
-#endif
 #endif
 }
 

--- a/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
+++ b/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
@@ -93,7 +93,6 @@ void SecureSocketImpl::acceptSSL()
 		throw SSLException("Cannot create SSL object");
 	}
 
-#if OPENSSL_VERSION_NUMBER >= 0x1010100fL
 	/* TLS 1.3 server sends session tickets after a handhake as part of
 	* the SSL_accept(). If a client finishes all its job before server
 	* sends the tickets, SSL_accept() fails with EPIPE errno. Since we
@@ -105,7 +104,6 @@ void SecureSocketImpl::acceptSSL()
 		throw SSLException("Cannot create SSL object");
 	}
 	//Otherwise we can perform two-way shutdown. Client must call SSL_read() before the final SSL_shutdown().
-#endif
 
 	::SSL_set_bio(_pSSL, pBIO, pBIO);
 	::SSL_set_accept_state(_pSSL);
@@ -178,12 +176,10 @@ void SecureSocketImpl::connectSSL(bool performHandshake)
 		SSL_set_tlsext_host_name(_pSSL, _peerHostName.c_str());
 	}
 
-#if OPENSSL_VERSION_NUMBER >= 0x10001000L
 	if(_pContext->ocspStaplingResponseVerificationEnabled())
 	{
 		SSL_set_tlsext_status_type(_pSSL, TLSEXT_STATUSTYPE_ocsp);
 	}
-#endif
 
 	if (_pSession && _pSession->isResumable())
 	{

